### ISSUE TITLE
Feature Request: Exit Level Option

### ISSUE THREAD

**Issue by ChristianSilvermoon** [Wohlstand/TheXTech#53]
The option to press **Start** and then **Drop Item** or simply an **Exit Level** option in the pause menu would be an awesome addition.

**Comment by 0lhi** [comment]
Good idea! I'd prefer Start + Drop Item to keep the Pause Menu consistent with other SMBX versions.

**Comment by Wohlstand** [comment]
The only thing I should to do: return player into the same state as before entering the level: I.e. to keep obtained power-up and updated progress, player must quit the level by normal way

**Comment by 0lhi** [comment]
In Super Mario World, the progress will be updated. But you can only Start + Select after you have beaten the level already.

**Comment by Wohlstand** [comment]
> In Super Mario World, the progress will be updated. But you can only Start + Select after you have beaten the level already.

Interesting, I.e. allow the level quit with progress keeping in only condition level got beaten before, and that won't work if level didn't beated before (you can't exist without progress loss until you'll beat this level). All I understood correcly? Anyway, I'll need to try the game at my home...

**Comment by Wohlstand** [comment]
btw, what about levels that has multiple stars? (you will need to replay the level for multiple times, the big example at "The Great Castle Adventure if I remind correctly") Or levels with a secret exit(s)?

**Comment by 0lhi** [comment]
> (you can't exist without progress loss until you'll beat his level)

Exiting won't work at all when the level wasn't beaten.

I have made two quick recordings: https://imgur.com/a/C0wFOyq (A bit buggy but it shows what's important.)

**Comment by Wohlstand** [comment]
> I have made two quick recordings: https://imgur.com/a/C0wFOyq (A bit buggy but it shows what's important.)

Damn Cinnamon + NVidia, however, I do use OBS in the NVENC mode, and it makes non-glitchy recordings a tleast

**Comment by 0lhi** [comment]
> Damn Cinnamon + NVidia, however, I do use OBS in the NVENC mode, and it makes non-glitchy recordings at least

Yeah :/ OBS does a better job for me too, but vokoscreen is more convenient for quick things.

Anyway, just one Exit is fine, not matter which one: https://imgur.com/a/fM84agi

**Comment by Wohlstand** [comment]
I see, so, any exit will allow this.

**Comment by Wohlstand** [comment]
This option will be added into "compat.ini" and will be unavailable at speed-run mode 2 and 3 (only mode 1) to give the fair result of speed runs that usually performing without this feature (also, this feature will give the biggest time winning, and therefore is should not allowed for sleed-runs, but allowed for casual playthroughs and speed-run mode 1 at TheXTech-targeted episodes)

**Comment by 0lhi** [comment]
Oh, btw! Checkpoints also get saved: https://imgur.com/a/pVlslVq

**Comment by Wohlstand** [comment]
wow

**Comment by Wohlstand** [comment]
Another thing on dealing with tricky-designed HUB episodes like the "Mario & Luigi Marathon" episode? btw, this episode would also get the "sub-hubs" option to resume the game at the level got entered by player last time by the same warp

**Comment by 0lhi** [comment]
It has been a while since I have played the Marathon. I think, episodes like that where the Hubworld is setup in a really special way, should have stuff like Start + Select or Subhubs disabled. I haven't played Mario Prime, but that also seems like an episode someone should take a look at.

Anyway, for the Sub-Hubs, the best episode to implement that first would be "Another Adventure".

I also think that compat features should be divided between those that only affect a given level (most bug-fixes), and those that affect the episode as a whole (Last-Hub Resume, Sub-Hubs, Start+Select). And for uniquely setup episodes such as Marathon or Prime, disable those episode-wide features.

**Comment by ds-sloth** [comment]
I spoke with Krissy about this today, and we came to the following idea:
- When a player hasn't beaten a level, it might still be important to escape the level (ie, softlock when the player hasn't saved for some time). In this case, provide a level exit identical to a death (costing one life and setting all players' states to State 1, but not resetting checkpoints), that doesn't activate the level restart mechanism if it's enabled for the episode.
- When a player *has* beaten a level, allow them to return directly to the world map, keeping their current state, checkpoints, and any collected powerups / stars, but not treating the level as completed for the purpose of Medal tracking. 
- In both cases, use the Select button at the pause menu. Require holding the button for around 1 second, and play a sound and display a message at the beginning of this period to alert the player.
- This will require adding level exit tracking (also needed for #746) to the LevelSaveInfo system. It should be easy to do this with a `uint16_t` bitfield.
- This will nerf "locked room challenges" such as the one in the Fungi Village in Cliche.

How does this sound to the two of you, @Wohlstand, @0lhi ?

**Comment by ChristianSilvermoon** [comment]
I believe for where to place the player the current logic of loading a save with sub hubs should be used, if and when needed, since ThexTech already handles a case like that

**Comment by Wohlstand** [comment]
I remembered the case of "Super Mario 3D World": player is able to quit the currently playing via menu with lose all the coins / stars / stamps / points / obtained/lost powerups played on this level. However, this option is active only in the case when all the players are standing on the hard ground (it's impossible to hit that option if player in a mid air).

**Comment by ChristianSilvermoon** [comment]
> I remembered the case of "Super Mario 3D World": player is able to quit the currently playing via menu with lose all the coins / stars / stamps / points / obtained/lost powerups played on this level. However, this option is active only in the case when all the players are standing on the hard ground (it's impossible to hit that option if player in a mid air).

Another game with such a feature is SM64, which forbids exiting the level if you are moving or not on solid ground

**Comment by ds-sloth** [comment]
I think that even if the player is not grounded, it should be possible for them to leave the level in the death-like manner.

But I would be fine with requiring the player to be grounded in the case where they've beaten the level.

**Comment by Superbloxen** [comment]
> I think that even if the player is not grounded, it should be possible for them to leave the level in the death-like manner.
> 
> But I would be fine with requiring the player to be grounded in the case where they've beaten the level.

What do you think about punishing the player by causing them to lose 50 coins or 1 stack of centicoins if they have not yet beaten the level once?

**Comment by ds-sloth** [comment]
Please refer to this comment which I think is still the plan for how this feature will be implemented: https://github.com/TheXTech/TheXTech/issues/53#issuecomment-2185468033

Losing 1 stack of coins is the same as leaving the level in a death-like manner, right?

**Comment by ChristianSilvermoon** [comment]
> Losing 1 stack of coins is the same as leaving the level in a death-like manner, right?

I would personally say yes
